### PR TITLE
enable peak memory measurement by default

### PR DESCRIPTION
**Pull request by FindHao**
The CPU and GPU peak memory measurements will be enabled by default.
The output is like the following:
```
$ python3 run.py -d cuda -t train   BERT_pytorch 
Running train method from BERT_pytorch on cuda in eager mode with input batch size 16.
GPU Time:            107.128 milliseconds
CPU Total Wall Time: 107.156 milliseconds
GPU 0 Peak Memory:                6.2205 GB
CPU Peak Memory:                2.8018 GB
```
The dependent package could be installed by the following command.
```
pip install pynvml
```

### CODE REVIEW

**Comment by xuzhao9** [comment]
Let's add `nvidia-ml-py` to requirements.txt in this case.

**Comment by xuzhao9** [comment]
Also, let's remove the logs 
```
Metric gpu_peak_mem is collected by nvml backend
Metric cpu_peak_mem is collected by psutil.Process.
```

**Comment by xuzhao9** [comment]
@FindHao One question: can we measure the memory usage in different runs than latency measurement loop? This makes sure memory and latency measurements are independent.

**Comment by xuzhao9** [comment]
> Do you think it is a good idea to use another [while loop](https://github.com/pytorch/benchmark/pull/1392/files#diff-d6af0459a37d985953d7040c14f53feb3b9cc9e58b543aa3c2b80256d276c5e0R131) to measure the memory usage?
Yes another while loop sounds good to me. Note that we don't need to warmup for memory usage.

> 
> What about TFLOPS?

TFLOPS is fine since it is not enabled by default.

**Comment by FindHao** [comment]
There are two solutions to split the memory measurements.
1. use another independent model_analyzer
2. use the existing model_analyzer

The first one is easier to be implemented, but the code in run.py could be slightly ugly. It has more overhead although this overhead is not counted in our measurements. The second one needs more changes in ModelAnalyzer, but the code in run.py is neater. Which one do you think is better?

**Comment by xuzhao9** [comment]
@FindHao Let's keep a neater run.py

File: torchbenchmark/util/experiment/metrics.py [hunk 1]
```diff
@@ -38,19 +40,58 @@ def get_latencies(func, device: str, nwarmup=WARMUP_ROUNDS, num_iter=BENCHMARK_I
...
+    if 'gpu_peak_mem' in metrics_needed:
+        gpu_peak_mem = mem_model_analyzer.calculate_gpu_peak_mem()
+    else:
+        gpu_peak_mem = None
+    if 'cpu_peak_mem' in metrics_needed:
+        cpu_peak_mem = mem_model_analyzer.calculate_cpu_peak_mem()
+    else:
+        cpu_peak_mem = None
+    if export_metrics_file:
+        mem_model_analyzer.update_export_name("_peak_memory")
```
Comment (xuzhao9) [inline torchbenchmark/util/experiment/metrics.py:69]: What is this metrics file?

File: torchbenchmark/util/experiment/metrics.py [hunk 2]
```diff
@@ -15,6 +15,8 @@
 @dataclasses.dataclass
 class TorchBenchModelMetrics:
     latencies: List[float]
+    cpu_peak_mem: float
```
Comment (xuzhao9) [inline torchbenchmark/util/experiment/metrics.py:18]: It should be Optional[float], not float.

File: torchbenchmark/util/experiment/metrics.py (hunk 1 above)
Comment (FindHao) [reply torchbenchmark/util/experiment/metrics.py:69]: `--export-metrics` in `run.py` will export all records to help us debug or find more detailed information.

File: torchbenchmark/util/experiment/metrics.py (hunk 2 above)
Comment (FindHao) [reply torchbenchmark/util/experiment/metrics.py]: fixed by https://github.com/pytorch/benchmark/pull/1392/commits/57f0649d86e35507ceb9bc1608e1f588744faa90

File: torchbenchmark/util/experiment/metrics.py [hunk 3]
```diff
@@ -38,19 +40,58 @@ def get_latencies(func, device: str, nwarmup=WARMUP_ROUNDS, num_iter=BENCHMARK_I
         result_summary.append((t1 - t0) / NANOSECONDS_PER_MILLISECONDS)
     return result_summary
 
+def get_peak_memory(func, device: str, num_iter=BENCHMARK_ITERS, export_metrics_file='', metrics_needed=[], metrics_gpu_backend='dcgm') -> Tuple[float, float]:
```
Comment (xuzhao9) [inline torchbenchmark/util/experiment/metrics.py:43]: I guess we don't need to run 15 iterations to get the memory metrics?
Since we plan to print this metrics by default, it is better to save some time.

File: requirements.txt [hunk 4]
```diff
@@ -20,3 +20,4 @@ numpy==1.21.2
 git+https://github.com/kornia/kornia.git@b7050c3
 scipy # for lazy_bench.py
 submitit
+nvidia-ml-py
```
Comment (xuzhao9) [inline requirements.txt:23]: Should we make it pynvml to make it consistent with the internal version?

File: components/model_analyzer/TorchBenchAnalyzer.py [hunk 5]
```diff
@@ -80,16 +99,26 @@ def set_gpu_monitor_backend_nvml(self):
         self.gpu_monitor_backend = 'nvml'
 
     def set_export_csv_name(self, export_csv_name=''):
+        if export_csv_name == '':
```
Comment (xuzhao9) [inline components/model_analyzer/TorchBenchAnalyzer.py:102]:
```suggestion
        if not export_csv_name:
```

File: components/model_analyzer/TorchBenchAnalyzer.py [hunk 6]
```diff
@@ -80,16 +99,26 @@ def set_gpu_monitor_backend_nvml(self):
...
+            return
         self.export_csv_name = export_csv_name
         # test for correct permission
         with open(export_csv_name, 'w') as fout:
             fout.write('')
 
+    def update_export_name(self, insert_str=''):
+        index = self.export_csv_name.find('.csv')
+        if index != -1:
```
Comment (xuzhao9) [inline components/model_analyzer/TorchBenchAnalyzer.py:111]:
```suggestion
        if not index == -1:
```

File: run.py [hunk 7]
```diff
@@ -321,6 +300,9 @@ def _validate_profile_options(profile_options: str):
     if args.amp:
         test = torch.autocast("cuda")(test)
     metrics_needed = [_ for _ in args.metrics.split(',') if _.strip()] if args.metrics else []
+    # enable cpu_peak_mem and gpu_peak_mem by default
+    metrics_needed.extend(['cpu_peak_mem', 'gpu_peak_mem'])
```
Comment (xuzhao9) [inline run.py:304]: Check if device is not `CUDA`, only collect `gpu_peak_mem`.

File: run.py [hunk 8]
```diff
@@ -340,7 +322,7 @@ def _validate_profile_options(profile_options: str):
             exit(-1)
         export_metrics_file = "%s_all_metrics.csv" % args.model
     else:
-        export_metrics_file = False
+        export_metrics_file = ''
```
Comment (xuzhao9) [inline run.py:325]: Can we use `None` here?

File: torchbenchmark/util/experiment/metrics.py (hunk 3 above)
Comment (FindHao) [reply torchbenchmark/util/experiment/metrics.py]: I use a new variable `MEMPROF_ITER` and set it to 2. It's better to set a bigger number like 4 if we want to get more stable and accurate results since some models' execution times are too short.

**Comment by FindHao** [comment]
thanks for your suggestions! the last commit has been fixed most of them. the remaining 2 issues are 
1.  I use a new variable MEMPROF_ITER and set it to 2. Do you think it is ok for now? It's better to set a bigger number like 4 if we want to get more stable and accurate results since some models' execution times are too short. 
2. I found [pynvml](https://pypi.org/project/pynvml/#history) is more like one version of [nvidia-ml-py](https://pypi.org/project/nvidia-ml-py/#history). Which one do you think is better to use? nvidia-ml-py is more actively developed but the package name when we try to import in python is still `pynvml` which is a little bit confusing.

File: torchbenchmark/util/experiment/metrics.py [hunk 9]
```diff
@@ -38,19 +41,59 @@ def get_latencies(func, device: str, nwarmup=WARMUP_ROUNDS, num_iter=BENCHMARK_I
         result_summary.append((t1 - t0) / NANOSECONDS_PER_MILLISECONDS)
     return result_summary
 
+def get_peak_memory(func, device: str, num_iter=MEMPROF_ITER, export_metrics_file='', metrics_needed=[], metrics_gpu_backend='dcgm') -> Tuple[float, float]:
```
Comment (xuzhao9) [inline torchbenchmark/util/experiment/metrics.py:44]: The return value is `Tuple[float, Optional[str], float]`

**Comment by xuzhao9** [comment]
> thanks for your suggestions! the last commit has been fixed most of them. the remaining 2 issues are
> 
> 1. I use a new variable MEMPROF_ITER and set it to 2. Do you think it is ok for now? It's better to set a bigger number like 4 if we want to get more stable and accurate results since some models' execution times are too short.

Is it possible to run 2 iterations and check the result is stable and accurate, if not, keep running at a maximum of 15 iterations?

**Comment by FindHao** [comment]
> Is it possible to run 2 iterations and check the result is stable and accurate, if not, keep running at a maximum of 15 iterations?

The problem is that several model's execution times are too short. model analyzer requests metric measurement records from DCGM backend by the specified interval. By default, model analyzer's frequency is every 10ms. But for DCGM backend, the frequency is about every 1.5ms. It means model analyzer will get about 6~7 records for each request. for short models, not all records has valid data. e.g., if the model execution time is 5ms, we only need the first 3 records. so I add [a timestamp](https://github.com/pytorch/benchmark/pull/1366) when the monitor stops. 

But, sometimes the DCGM timestamps are all larger than the stop timestamp, we have to keep [at least one record](https://github.com/pytorch/benchmark/pull/1391). It is the extreme case. 

DCGM's monitoring crosses multiple iterations, and actually we don't know the execution time for each iteration when model analyzer collects metrics. So it could be hard to check the results' stableness or accuracy online.
But if we have more iterations, we will get more accurate results. I think 2 iterations should be OK for us, even for short models. I'll do the accuracy tests offline to double check it.

**Comment by xuzhao9** [comment]
Another idea is to keep running the model for some time threshold (say 100ms) if single iteration is too short, and if single iteration is long enough, just run 1 or 2 iterations minimum.

**Comment by FindHao** [comment]
dcgm monitor starts before the iterations, ends after all iterations, and works on another thread. So it is hard to let dcgm monitor know each iteration's execution time. If we move the monitors' start and end inside the loop, there would be much more overhead, and it would be complicated to handle the collected data. Since the sampling granularity of dcgm is about 1.5ms, if we make it inside the loop, there would be more noise data.

**Comment by FindHao** [comment]
A possible way is that after the 2 iterations end, we check the execution time. if it is too short, run another memory measurement with more iterations.

**Comment by xuzhao9** [comment]
> A possible way is that after the default 2 iterations end, we check the execution time. if it is too short, run another memory measurement with more iterations.

Yes, that sounds good to me. The basic idea is that we don't need to measure the iteration execution time within the ModelAnalzyer.

**Comment by xuzhao9** [comment]
I get this failure when running internally:

```
  File "/mnt/xarfuse/uid-234232/7d206292-seed-nspid4026531836_cgpid114977510-ns-4026531840/pytorch/benchmark/run.py", line 158, in run_one_step
    cpu_peak_mem, device_id, gpu_peak_mem = get_peak_memory(func, model.device, export_metrics_file=export_metrics_file, metrics_needed=metrics_needed, metrics_gpu_backend=metrics_gpu_backend)
  File "/mnt/xarfuse/uid-234232/7d206292-seed-nspid4026531836_cgpid114977510-ns-4026531840/torchbenchmark/util/experiment/metrics.py", line 70, in get_peak_memory
    mem_model_analyzer.start_monitor()
  File "/mnt/xarfuse/uid-234232/7d206292-seed-nspid4026531836_cgpid114977510-ns-4026531840/components/model_analyzer/TorchBenchAnalyzer.py", line 118, in start_monitor
    self.gpu_factory = GPUDeviceFactory(self.gpu_monitor_backend)
  File "/mnt/xarfuse/uid-234232/7d206292-seed-nspid4026531836_cgpid114977510-ns-4026531840/components/model_analyzer/tb_dcgm_types/gpu_device_factory.py", line 38, in __init__
    self.init_all_devices()
  File "/mnt/xarfuse/uid-234232/7d206292-seed-nspid4026531836_cgpid114977510-ns-4026531840/components/model_analyzer/tb_dcgm_types/gpu_device_factory.py", line 94, in init_all_devices
    gpu_device = GPUDevice(device_name, device_id, pci_bus_id, device_uuid)
  File "/mnt/xarfuse/uid-234232/7d206292-seed-nspid4026531836_cgpid114977510-ns-4026531840/components/model_analyzer/tb_dcgm_types/gpu_device.py", line 50, in __init__
    assert type(device_name) is str
AssertionError
```

File: components/model_analyzer/tb_dcgm_types/gpu_device_factory.py [hunk 10]
```diff
@@ -47,34 +48,58 @@ def init_all_devices(self, dcgmPath=None):
...
+            import pynvml
+
+            logger.info("Initializing GPUDevice handles using NVML")
+            pynvml.nvmlInit()
+
+            # Create a GPU device for every supported NVML device
+            nvml_device_count = pynvml.nvmlDeviceGetCount()
+            for device_id in range(nvml_device_count):
+                handle = pynvml.nvmlDeviceGetHandleByIndex(device_id)
+                device_name = pynvml.nvmlDeviceGetName(handle)
```
Comment (xuzhao9) [inline components/model_analyzer/tb_dcgm_types/gpu_device_factory.py:90]: We need to decode the name here
```suggestion
                device_name = pynvml.nvmlDeviceGetName(handle).decode('ascii')
                pci_bus_id = pynvml.nvmlDeviceGetPciInfo(handle).busId.decode('ascii')
                device_uuid = pynvml.nvmlDeviceGetUUID(handle).decode('ascii')
```

Otherwise they won't pass the check at https://github.com/pytorch/benchmark/blob/main/components/model_analyzer/tb_dcgm_types/gpu_device.py#L51

Comment (FindHao) [reply components/model_analyzer/tb_dcgm_types/gpu_device_factory.py:90]: Sorry. I should do more compatibility checks. Testing on pynvml 11.4.1 now. Will fix them soon.

**Comment by xuzhao9** [comment]
Also I found a TODO here: https://github.com/pytorch/benchmark/blob/main/components/model_analyzer/tb_dcgm_types/gpu_device.py#L16 
Maybe we can also migrate to nvml and deprecate numba?

**Comment by FindHao** [comment]
> 

This error has been set to `info` level. Bye default, the logger level in model analyzer is `warning`. So this message will not be shown unless we manually change the logger level. 

numba could detect all GPUs even if you use `NVIDIA_VISIBLE_DEVICES` to limit which GPU could be visible. I think it's better to leave it here for debugging purposes.

**Comment by xuzhao9** [comment]
>  I think it's better to leave it here for debugging purposes. But let me double check if we still need it after we migrate to pynvml.

Let's still remove this, otherwise people will complain for noisy logs.

**Comment by FindHao** [comment]
> > I think it's better to leave it here for debugging purposes. But let me double check if we still need it after we migrate to pynvml.
> 
> Let's still remove this or manually move this to debug level, otherwise people will complain for noisy logs.

Sure. will fix it in the next commit.

**Comment by FindHao** [comment]
@xuzhao9 the warning messages have been commented. It passed the test on pynvml 11.4.1. 
But I failed to get rid of numba. Some information like sm count  is not in pynvml.

**Comment by xuzhao9** [comment]
@FindHao That's fine, we still need sm counts so let's keep numba.

**Comment by xuzhao9** [comment]
@FindHao Can you also move the following logs to debug? We don't need them as INFO.
```
INFO:model_analyzer_logger:Initializing GPUDevice handles using NVML
INFO:numba.cuda.cudadrv.driver:init
INFO:model_analyzer_logger:Using GPU 0 NVIDIA AAA with UUID GPU-BBB
```
